packages/convex-helpers/server/retries.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * This file defines helper functions that can be used to retry a
+ * Convex action until it succeeds. An action should only be retried if it is
+ * safe to do so, i.e., if it's idempotent or doesn't have any unsafe side effects.
+ */
+import {
+  FunctionReference,
+  FunctionVisibility,
+  Scheduler,
+  FunctionArgs,
+  OptionalRestArgs,
+  getFunctionName,
+  makeFunctionReference,
+  DefaultFunctionArgs,
+  internalMutationGeneric,
+} from "convex/server";
+import { v, ObjectType } from "convex/values";
+
+const DEFAULT_WAIT_BACKOFF = 10;
+const DEFAULT_RETRY_BACKOFF = 10;
+const DEFAULT_BASE = 2;
+const DEFAULT_MAX_FAILURES = 16;
+
+/**
+ * Create a function that retries an action with exponential backoff.
+ * e.g.
+ * ```ts
+ * // in convex/utils.ts
+ * import { makeActionRetrier } from "convex-helpers/server/retries";
+ *
+ * export const { runWithRetries, retry } = makeActionRetrier("utils:retry");
+ *
+ * // in a mutation or action
+ * await runWithRetries(ctx, internal.myModule.myAction, { arg1: 123 });
+ * ```
+ *
+ * @param internalMutation From "./convex/_generated/server" or customMutation.
+ * @param retryRef The function reference to the retryRef function exported.
+ * e.g. internal.mymodule.retry
+ * @returns An object with runWithRetries and retry functions.
+ */
+export function makeActionRetrier(retryFnName: string) {
+  const retryRef = makeFunctionReference<
+    "action",
+    ObjectType<typeof retryArguments>
+  >(retryFnName);
+  /**
+   * Run and retry action until it succeeds or fails too many times.
+   *
+   * @param action - Name of the action to run, e.g., `usercode:maybeAction`.
+   * @param actionArgs - Arguments to pass to the action, e.g., `{"failureRate": 0.75}`.
+   * @param options
+   * @param options.waitBackoff=DEFAULT_WAIT_BACKOFF (10) - Initial delay before checking action status, in milliseconds.
+   * @param options.retryBackoff=DEFAULT_RETRY_BACKOFF (10) - Initial delay before retrying, in milliseconds.
+   * @param options.base=DEFAULT_BASE (2) - Base of the exponential backoff.
+   * @param options.maxFailures=DEFAULT_MAX_FAILURES (16) - The maximum number of times to retry the action.
+   */
+  async function runWithRetries<
+    Action extends FunctionReference<
+      "action",
+      Visibility,
+      DefaultFunctionArgs,
+      null | Promise<null> | void | Promise<void>
+    >,
+    Visibility extends FunctionVisibility = "internal"
+  >(
+    ctx: { scheduler: Scheduler },
+    action: Action,
+    actionArgs: FunctionArgs<Action>,
+    options?: {
+      waitBackoff?: number;
+      retryBackoff?: number;
+      base?: number;
+      maxFailures?: number;
+    }
+  ) {
+    const job = await ctx.scheduler.runAfter(
+      0,
+      action,
+      ...([actionArgs] as OptionalRestArgs<Action>)
+    );
+    await ctx.scheduler.runAfter(0, retryRef, {
```

Comment:
If the caller fails at this point the action won't be reliable right?

I think we can't put out a framework for reliable actions that isn't actually reliable. Seems actually quite plausible that the action would fail right here. Is there a reason why we can't have the one scheduled mutation both call the action and retry it?

Reply:
tl;dr I think I'll change it to also require a `db: DatabaseWriter` in the ctx, so that it's only callable from mutations to encourage the right behavior. it can be wrapped to call from an action / client.

My thought was that we should be encouraging folks to do this from the server, not expose a public action - public actions can be client-side retried. Or wrap with a mutation.
My reasoning for not putting in its own internalMutation:
A: We don't suggest directly calling mutations from each other, so it won't be transactional with your other data changes when scheduled from a mutation.

B: In an action here, it can fail:

1. before scheduling anything - action will be attempted 0 times
2. between scheduling the action and scheduling the retry - action will be attempted 1 time
3. after scheduling - action will be attempted N times, but further logic won't be attempted.

When you schedule a mutation from an action, it can fail:

1. before scheduling anything - action will be attempted 0 times
2. after scheduling - action will be attempted N times, but further logic won't be attempted.

The incremental "loss" in an action is the possibility that you'd only try something once. 

If we want to avoid this, we should push users to use it transactionally with their other data commits, and force them to wrap it explicitly to call from an action.

Down the line I'd like to roll this sort of behavior into a job mgmt helper where you can subscribe to the job status - how many times its been retried etc. right now you get an ID that is only for the first invocation and can't trace its progress

Reply:
I've edited the above, and also realized that since the only "transactional" benefit here is that they are scheduled, and scheduling the wrapper is transactional, there's no transactional benefit. 

**The main reason is that this version has type checking. But we can constrain it to being called by mutations if we want to call it but also ensure it's in a mutation**

Reply:
I think we shouldn't give the developer an option to ever call this directly from an action. The source code in here is complicated enough that they likely won't really know what's going on and instead will just use it as a black box. So I think we just give them explicit examples of what to do when you want to call a reliable action either from the client, from a mutation, or from another action, the latter which will require scheduling through a mutation.

I think we'll be able to ship once we just are very prescriptive with the example instead of giving them options. This could be as simple as a comment similar to what we have but says explicitly something like "don't call this from an action. schedule a mutation that calls this"